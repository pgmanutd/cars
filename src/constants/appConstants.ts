export const API_BASE_PATH = '/api';

export const QUERY_KEYS = {
  sort: 'sort',
  page: 'page',
  colorFilter: 'filters.color',
  manufacturerFilter: 'filters.manufacturer',
};

export const LOCAL_STORAGE_KEYS = {
  favoriteCars: 'favoriteCars',
  language: 'language',
};

export const DEFAULT_QUERY_VALUES = {
  sort: '',
  page: '1',
  colorFilter: '',
  manufacturerFilter: '',
};

export const DEFAULT_PARAM_VALUES = {
  stockNumber: '',
};
