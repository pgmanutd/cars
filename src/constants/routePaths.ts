import { StockNumberParam } from 'features/CarInfo/types';

const routePaths = {
  cars: '/',
  carDetails: {
    pathWithParams: ({ stockNumber }: { stockNumber: StockNumberParam }) =>
      `/details/${stockNumber}`,
    path: '/details/:stockNumber',
  },
  favorites: '/favorites',
  notFound: '/not-found',
};

export default routePaths;
