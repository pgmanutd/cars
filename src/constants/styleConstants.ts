export const HEADER_HEIGHT = 80;

export const FOOTER_HEIGHT = 80;

export const CONTAINER_MAX_WIDTH = 800;
