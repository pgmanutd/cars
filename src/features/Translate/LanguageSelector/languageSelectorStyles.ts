import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    formControl: {
      width: '100%',
    },
    selectEmpty: {
      marginTop: theme.spacing(3),
    },
  }),
);
