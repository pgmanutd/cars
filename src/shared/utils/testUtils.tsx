import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';
import { render } from '@testing-library/react';

export const renderWithRouter = (
  ui: React.ReactElement,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  }: { route?: string; history?: MemoryHistory } = {},
) => ({
  ...render(<Router history={history}>{ui}</Router>),
  history,
});
