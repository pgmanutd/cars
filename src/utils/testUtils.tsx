import React from 'react';
import { Router, Route } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';
import { HelmetProvider } from 'react-helmet-async';
import { render } from '@testing-library/react';

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    routerConfig: {
      path = '/',
      route = '/',
      history = createMemoryHistory({ initialEntries: [route] }),
    },
  }: {
    routerConfig: { path?: string; route?: string; history?: MemoryHistory };
  } = { routerConfig: {} },
) => ({
  ...render(
    <HelmetProvider>
      <Router history={history}>
        <Route
          path={path}
          render={routeProps => React.cloneElement(ui, { ...routeProps })}
        />
      </Router>
    </HelmetProvider>,
  ),
  history,
});
