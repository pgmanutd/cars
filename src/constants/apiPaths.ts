import { API_BASE_PATH } from 'constants/appConstants';

import appendBasePath from 'utils/appendBasePath';

import { StockNumberParam } from 'features/CarInfo/types';
import { ManufacturerQuery, ColorQuery } from 'features/NavFilter/types';
import { PageQuery } from 'features/CarList/types';
import { SortQuery } from 'features/MileageSorter/types';

const apiPaths = {
  cars: (params: {
    manufacturer: ManufacturerQuery;
    color: ColorQuery;
    sort: SortQuery;
    page: PageQuery;
  }) =>
    appendBasePath(
      `${API_BASE_PATH}/cars`,
      new URLSearchParams(params)?.toString(),
    ),
  carDetails: ({ stockNumber }: { stockNumber: StockNumberParam }) =>
    `${API_BASE_PATH}/cars/${stockNumber}`,
  colors: () => `${API_BASE_PATH}/colors`,
  manufacturers: () => `${API_BASE_PATH}/manufacturers`,
};

export default apiPaths;
