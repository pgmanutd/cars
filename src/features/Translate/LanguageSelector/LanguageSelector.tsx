import React, { memo, useCallback } from 'react';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';

import pickTargetValue from 'utils/pickTargetValue';

import { Language } from '../types';
import useTranslate from '../useTranslate';

import { useStyles } from './languageSelectorStyles';

const LanguageSelector: React.FC = props => {
  const classes = useStyles();
  const { translate, language, languages, setLanguage } = useTranslate();

  const handleLanguageChange = useCallback(
    (event: React.ChangeEvent<{ value: unknown }>) => {
      setLanguage(pickTargetValue(event) as Language);
    },
    [setLanguage],
  );

  return (
    <FormControl
      data-testid="LanguageSelector"
      {...props}
      className={classes.formControl}
    >
      <InputLabel shrink id="LanguageSelectorLabel">
        {translate('features.LanguageSelector.languageLabel')}
      </InputLabel>
      <Select
        labelId="LanguageSelectorLabel"
        id="LanguageSelectorLabelSelect"
        value={language}
        onChange={handleLanguageChange}
        displayEmpty
        className={classes.selectEmpty}
      >
        {languages.map(languageKey => (
          <MenuItem key={languageKey} value={languageKey}>
            <em>
              {translate<string>(
                `features.LanguageSelector.${languageKey}Language`,
              )}
            </em>
          </MenuItem>
        ))}
      </Select>
    </FormControl>
  );
};

export default memo(LanguageSelector);
