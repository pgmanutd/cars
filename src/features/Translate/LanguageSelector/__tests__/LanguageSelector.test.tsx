import React from 'react';
import userEvent from '@testing-library/user-event';

import { LOCAL_STORAGE_KEYS } from 'constants/appConstants';

import { renderWithTranslate } from 'utils/testUtils';

import LanguageSelector from '../LanguageSelector';

describe('<LanguageSelector />', () => {
  const setup = () => {
    const { renderResult } = renderWithTranslate(<LanguageSelector />);

    return {
      renderResult,
    };
  };

  it('should render the component and matches it against stored snapshot', () => {
    const { renderResult } = setup();

    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it('should add selected language to localstorage when "German" item in dropdown is clicked', () => {
    const { renderResult } = setup();

    userEvent.click(renderResult.getByText('English'));
    userEvent.click(renderResult.getByText('German'));

    expect(localStorage.getItem(LOCAL_STORAGE_KEYS.language)).toBe('"de"');
  });
});
