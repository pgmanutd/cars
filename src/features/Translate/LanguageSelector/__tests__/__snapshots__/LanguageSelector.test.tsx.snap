// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LanguageSelector /> should render the component and matches it against stored snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root makeStyles-formControl-1"
    data-testid="LanguageSelector"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink="true"
      id="LanguageSelectorLabel"
    >
      Language
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-selectEmpty-2 MuiInputBase-formControl MuiInput-formControl"
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="LanguageSelectorLabel LanguageSelectorLabelSelect"
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        id="LanguageSelectorLabelSelect"
        role="button"
        tabindex="0"
      >
        <em>
          English
        </em>
      </div>
      <input
        type="hidden"
        value="en"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
